commit 59f606e4c21959776175c1118f0067b4b6ab8c6d (refs/changes/15/1715/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-03-24T19:11:35+01:00 (2 years, 7 months ago)
    
    TOOLS-1728 sdcadm update for service with multiple instances should skip up-to-date ones

diff --git a/lib/procedures/index.js b/lib/procedures/index.js
index e07cadd..c685759 100644
--- a/lib/procedures/index.js
+++ b/lib/procedures/index.js
@@ -102,6 +102,7 @@ function coordinatePlan(opts, cb) {
     var log = opts.log;
     var progress = opts.progress || function () {};
     var sdcadm = opts.sdcadm;
+    var forceSameImage = opts.plan.forceSameImage;
     var instsFromSvcName = {};
     var insts = opts.plan.curr;
     for (var i = 0; i < insts.length; i++) {
@@ -216,7 +217,12 @@ function coordinatePlan(opts, cb) {
                             handle.push(change);
                         } else {
                             if (~HA_READY_SVCS.indexOf(change.service.name)) {
-                                change.insts = svcInsts;
+                                var chInsts = forceSameImage ? svcInsts :
+                                        svcInsts.filter(function (ins) {
+                                            return (ins.image !==
+                                                change.image.uuid);
+                                        });
+                                change.insts = chInsts;
                                 handle.push(change);
                             } else {
                                 log.debug({
@@ -381,7 +387,12 @@ function coordinatePlan(opts, cb) {
                 {
                     var svcInsts = instsFromSvcName[change.service.name] || [];
                     if (svcInsts.length && svcInsts.length > 1) {
-                        change.insts = svcInsts;
+                        var chInsts = forceSameImage ? svcInsts :
+                                svcInsts.filter(function (ins) {
+                                    return (ins.image !==
+                                        change.image.uuid);
+                                });
+                        change.insts = chInsts;
                     } else {
                         change.inst = svcInsts[0];
                     }
diff --git a/lib/procedures/update-binder-v2.js b/lib/procedures/update-binder-v2.js
index 4e130b8..bb7f6d9 100644
--- a/lib/procedures/update-binder-v2.js
+++ b/lib/procedures/update-binder-v2.js
@@ -247,6 +247,10 @@ UpdateBinderV2.prototype.execute = function ushiExecute(opts, cb) {
                 vasync.forEachPipeline({
                     inputs: followers,
                     func: function reprovFollower(inst, next_) {
+                        if (inst.image === change.image.uuid) {
+                            next_();
+                            return;
+                        }
                         s.reprovisionRemote({
                             server: inst.server,
                             img: change.image,
@@ -304,8 +308,9 @@ UpdateBinderV2.prototype.execute = function ushiExecute(opts, cb) {
             },
 
             function reprovisionLeader(_, next) {
-                if (!leader) {
-                    return next();
+                if (!leader || leader.image === change.image.uuid) {
+                    next();
+                    return;
                 }
                 progress('Updating ZK leader');
                 s.reprovisionRemote({
